atom/browser/net/atom_network_delegate.cc

```diff
@@ -0,0 +1,388 @@
+      header_dict->SetString(key, value);
+  }
+  return header_dict;
+}
+
+void OnBeforeURLRequestResponse(
+    const net::CompletionCallback& callback,
+    GURL* new_url,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (!result.redirectURL.is_empty())
+    *new_url = result.redirectURL;
+  callback.Run(result.Cancel());
+}
+
+void OnBeforeSendHeadersResponse(
+    const net::CompletionCallback& callback,
+    net::HttpRequestHeaders* headers,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (!result.requestHeaders.IsEmpty())
+    *headers = result.requestHeaders;
+  callback.Run(result.Cancel());
+}
+
+void OnHeadersReceivedResponse(
+    const net::CompletionCallback& callback,
+    const net::HttpResponseHeaders* original_response_headers,
+    scoped_refptr<net::HttpResponseHeaders>* override_response_headers,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (result.responseHeaders.get()) {
+    *override_response_headers = new net::HttpResponseHeaders(
+        original_response_headers->raw_headers());
+    void* iter = nullptr;
+    std::string key;
+    std::string value;
+    while (result.responseHeaders->EnumerateHeaderLines(&iter, &key, &value)) {
+      (*override_response_headers)->RemoveHeader(key);
+      (*override_response_headers)->AddHeader(key + ": " + value);
+    }
+  }
+  callback.Run(result.Cancel());
+}
+
+}  // namespace
+
+// static
+std::map<AtomNetworkDelegate::EventTypes, AtomNetworkDelegate::ListenerInfo>
+AtomNetworkDelegate::event_listener_map_;
+
+AtomNetworkDelegate::AtomNetworkDelegate() {
+}
+
+AtomNetworkDelegate::~AtomNetworkDelegate() {
+}
+
+void AtomNetworkDelegate::SetListenerInIO(
+    EventTypes type,
+    scoped_ptr<base::DictionaryValue> filter,
+    const Listener& callback) {
+  ListenerInfo info;
+  info.callback = callback;
+
+  const base::ListValue* url_list = nullptr;
+  if (filter->GetList("urls", &url_list)) {
+    for (size_t i = 0; i < url_list->GetSize(); ++i) {
+      std::string url;
+      extensions::URLPattern pattern;
+      if (url_list->GetString(i, &url) &&
+          pattern.Parse(url) == extensions::URLPattern::PARSE_SUCCESS) {
+        info.url_patterns.insert(pattern);
+      }
+    }
+  }
+
+  event_listener_map_[type] = info;
+}
+
+int AtomNetworkDelegate::OnBeforeURLRequest(
+    net::URLRequest* request,
+    const net::CompletionCallback& callback,
+    GURL* new_url) {
+  brightray::NetworkDelegate::OnBeforeURLRequest(request, callback, new_url);
+
+  auto listener_info = event_listener_map_[kOnBeforeRequest];
+
+  if (!MatchesFilterCondition(request, listener_info))
+    return net::OK;
+
+  if (!listener_info.callback.is_null()) {
+    auto wrapped_callback = listener_info.callback;
+    auto details = ExtractRequestInfo(request);
+
+    BrowserThread::PostTaskAndReplyWithResult(BrowserThread::UI, FROM_HERE,
+        base::Bind(&RunListener, wrapped_callback, base::Passed(&details)),
```

Comment:
Nope, compiier reports `error: field of type 'base::DictionaryValue' has private copy constructor`